### PR TITLE
refactor!: Split out consumer genesis state

### PR DESCRIPTION
**Pull request by bermuell**
## Description

Closes: #1214

Refacoring shared data between provider and consumer.

Note: This breaks json format used by augmenting Genesis files of consumer chains with consumer genesis content exported from provider chain. Consumer Genesis content exported from a provider chain using major version 1,2 or3 of the provider module needs to be transformed with the transformation command introduced by this PR:

```
Transform the consumer genesis file from a provider version v1,v2 or v3 to a version supported by this consumer. Result is printed to STDOUT.

Example:
$ <appd> transform /path/to/ccv_consumer_genesis.json

Usage:
  interchain-security-cd genesis transform [genesis-file] [flags]
```

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if state-machine breaking change (i.e., requires coordinated upgrade)
* [x] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [x] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [x] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release for both the consumer and provider

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

**Comment by MSalopek** [comment]
The refactor looks good, thanks for the awesome work!

What is a bit unclear to me is if we should maintain backward compatibility with regards to responses from the `consumer-genesis` provider query:
```shell
interchain-security-pd query provider consumer-genesis <chain-id>
```

The structure returned by this query is an integral part of starting a consumer chain, so maybe we should address the issue of provider-consumer version compatibility.

This is the scenario I'm trying to solve:
1. consumer is on the newest version of ICS (ICS@latest)
2. provider is on an old version of ICS (lCS@v3.1)
3. after `spawn_time` the `consumer_genesis` becomes available on the provider
4. the consumer cannot parse the `consumer_genesis` in the format sent by the provider chain when doing `interchain-security-pd query provider consumer-genesis <chain-id>` since the provider is sending an old version of the genesis

It seems to me that this issue can be solved by parsing the "old" genesis into the "new" genesis:
```shell
interchain-security-cd parse-genesis v1 <path_to_v1_genesis>

# output of this cmd is a genesis that the consumer can process (ICS@latest)
```

If the provider and consumer are on the same ICS version there are no compatibility issues.

**Comment by bermuell** [comment]
thanks @MSalopek for your comment. 
Your scenario is the exactly the one we need to provide a way to deal with it.
There is already such a CLI (similar to what you suggested) to deal with that:
```
 interchain-security-cd genesis migrate -h
Migrate the source genesis into the target version and print to STDOUT.

Usage:
  interchain-security-cd genesis migrate [target-version] [genesis-file] [flags]
```

I'll look into that and provide migration code for my changes

**Comment by smarshall-spitzbart** [comment]
Matija's suggestion of creating a new parsing function sounds solid to me with appropriate docs for the procedure 👍

File: proto/interchain_security/ccv/consumer/v1/genesis.proto [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import "google/protobuf/timestamp.proto";
+
+
+
+// GenesisState defines the CCV consumer genesis state
+//
+// Note: this type is only used on consumer side and references shared types with
+// provider
+message GenesisState {
+  reserved 5, 6, 8;
```
Comment (smarshall-spitzbart) [inline proto/interchain_security/ccv/consumer/v1/genesis.proto:19]: What  is this `reserved` call used for ?

Comment (smarshall-spitzbart) [reply proto/interchain_security/ccv/consumer/v1/genesis.proto:19]: Got it, great stuff!

File: proto/interchain_security/ccv/consumer/v1/genesis.proto [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import "tendermint/abci/types.proto";
+
+
+
+
+// GenesisState defines the CCV consumer genesis state
+//
+// Note: this type is only used on consumer side and references shared types with
+// provider
+message GenesisState {
```
Comment (shaspitz) [inline proto/interchain_security/ccv/consumer/v1/genesis.proto:22]: In https://github.com/cosmos/interchain-security/pull/1288 we renamed the consumer's GenesisState struct to `ConsumerGenesisState`, is there a reason why we change it back in this PR?

File: app/consumer/genesis.go [hunk 3]
```diff
@@ -15,7 +28,130 @@ import (
...
+type (
+	// Transformation callback for a consumer genesis 'section' exported from a
+	// specific version of a provider
+	TransformationCallback func([]byte, client.Context) (json.RawMessage, error)
+
+	// TransformationMap defines the mapping from a version to a transformation callback
+	TransformationMap map[string]TransformationCallback
+)
+
+// Migration of consumer genesis content as it is exported from a provider version v2
```
Comment (shaspitz) [inline app/consumer/genesis.go:40]: Prob need some clarification on the use of "v2" and "v3" here, it seems like the term is applied to both consumer and provider. Are we referring to ICS major versions?

File: proto/interchain_security/ccv/consumer/v1/genesis.proto (hunk 2 above)
Comment (bermuell) [reply proto/interchain_security/ccv/consumer/v1/genesis.proto:22]: We changed it in the shared type namespace to make it clear that it's consumer related. This type is the result of the split and resides  in the consumer namespaces so no need to repeat that in the name. The shared one kept it's name

File: app/consumer/genesis.go (hunk 3 above)
Comment (bermuell) [reply app/consumer/genesis.go:40]: I was looking for a way how a user could find out what  module version is used on a provider chain for the provider implementation. I found the command `gaiad query upgrade module_versions` suitable for that. If you have another approach in mind pls let me know.
The above command returned the following on a real node:
```
...
{
  "name": "provider",
  "version": "2"
}
...
```

**Comment by MSalopek** [comment]
@bermuell docs are needed for the change to CLI.

File: app/consumer/genesis.go [hunk 4]
```diff
@@ -15,7 +28,130 @@ import (
...
+
+	newJson, err := ctx.Codec.MarshalJSON(&newGenesis)
+	if err != nil {
+		return nil, fmt.Errorf("failed marshalling data to new type: %s", err)
+	}
+	return newJson, nil
+}
+
+// Mapping of provider module version to related transformation function
+var transformationMap = TransformationMap{
```
Comment (MSalopek) [inline app/consumer/genesis.go:82]: I think we can remove this and simply comment what the CLI `transform` does.

The `transform` should only do the following:
* transform consumer genesis output coming from providers running ICSv1/2/3 into the latest consumer genesis version

**Review by MSalopek — APPROVED** [review]
LGTM!

File: app/consumer/genesis.go [hunk 5]
```diff
@@ -15,7 +26,96 @@ import (
...
 	return ModuleBasics.DefaultGenesis(cdc)
 }
+
+// GetConsumerGenesisTransformCmd transforms Consumer Genesis JSON content exported from a
+// provider version v1,v2 or v3 to a JSON format supported by this consumer version.
+func GetConsumerGenesisTransformCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "transform [genesis-file]",
+		Short: "Transform CCV consumer genesis from an older provider version not supporting current format",
+		Long: fmt.Sprintf(`Transform the consumer genesis file from a provider version v1,v2 or v3 to a version supported by this consumer. Result is printed to STDOUT.
```
Comment (sainoe) [inline app/consumer/genesis.go:112]: nit: perhaps add a `strings.TrimSpace()` call here

**Review by sainoe — APPROVED** [review]
LGTM! See my comment.

**Pull request merged**